library/std/src/collections/hash/map.rs

```diff
@@ -1151,6 +1151,28 @@ where
     }
 }
 
+#[stable(feature = "hashmap_from_array", since = "1.53.0")]
+impl<K, V, const N: usize> From<[(K, V); N]> for HashMap<K, V, RandomState>
+where
+    K: Eq + Hash,
+{
+    /// ```
+    /// use std::collections::HashMap;
+    /// let map1 = HashMap::from([(1, 2), (3, 4)]);
+    /// let map2: HashMap<_, _> = [(1, 2), (3, 4)].into();
+    /// assert_eq!(map1, map2);
+    /// let map3 = HashMap::from([(3, 4), (1, 2), (1, 2)]);
+    /// assert_eq!(map1, map3);
+    /// ```
+    fn from(arr: [(K, V); N]) -> Self {
+        let mut map = HashMap::with_capacity(N);
+        for (k, v) in crate::array::IntoIter::new(arr) {
```

Comment:
I'd just delegate to collect here, which is where all of our specialization-based optimizations tend to live.

Reply:
Sure thing. Some cursory examination of the assembly revealed that collect was doing "something" a bit differently than the `for` or `extend` implementations, but I couldn't tell if it was either better or worse.